File: ui/app/models/node.js
import DS from 'ember-data';
import { computed } from '@ember/object';

export default DS.Model.extend({
  name:           DS.attr('string'),
  category:       DS.attr('string'),
  module:         DS.attr('string'),
  config:         DS.attr(),
  info:           DS.attr('string'),
  isPrototype:    DS.attr('boolean'),
  x:              DS.attr('number'),
  y:              DS.attr('number'),
  ui_attributes:  DS.attr(),
  
  flowId:         DS.belongsTo('flow'),
  in_conns:       DS.hasMany('connections', {inverse: 'targetId'}),
  out_conns:      DS.hasMany('connections', {inverse: 'sourceId'}),

  jsPlumbEndpoints: DS.attr(),
  
  visibleName: computed("name", "config.name.value", function(){
    let config = this.get('config');

    if (config && config.name && config.name.value) {
      return config.name.value;
    } else {
      return this.get('name');
    }    
  })

});

File: apps/exred_ui/lib/exred_ui/editor/node.ex
defmodule ExredUI.Editor.Node do
  use Ecto.Schema
  import Ecto.Changeset
  alias ExredUI.Editor.Node

  @primary_key {:id, :binary_id, autogenerate: true}
  @derive {Phoenix.Param, key: :id}

  schema "nodes" do
    field :name, :string
    field :category, :string
    field :config, :map
    field :info, :string
    field :module, :string
    field :is_prototype, :boolean
    field :x, :integer
    field :y, :integer
    field :ui_attributes, :map

    belongs_to :flow, ExredUI.Editor.Flow, type: :binary_id
    has_many :in_conns, ExredUI.Editor.Connection, foreign_key: :target_id
    has_many :out_conns, ExredUI.Editor.Connection, foreign_key: :source_id

    timestamps()
  end

  @doc false
  def changeset(%Node{} = node, attrs) do
    node
    |> cast(attrs, [:id, :flow_id, :name, :module, :config, :info, :is_prototype, :category, :x, :y, :ui_attributes])
    |> validate_required([:id, :name, :module, :config, :info, :is_prototype, :category])
  end
end
